ocaml/xapi/xapi_cluster.ml

```diff
@@ -47,6 +47,50 @@ let create ~__context ~pIF ~cluster_stack ~pool_auto_join ~token_timeout ~token_
       assert_pif_prerequisites (pIF,pifrec);
       let ip = ip_of_pif (pIF,pifrec) in
 
+      let assert_pool_on_same_subnet () =
+        (** we gather ip addresses from all the hosts and also the netmask from the PIF.
+          * if all the ip addresses do not have the same network address, then we raise an
+            error *)
+        let are_on_same_subnet =
+            let parse_ip ip_str =
+              match Ipaddr.of_string ip_str with
+              | Ok ip_addr       -> ip_addr
+              (* this shouldn't happen because we shouldn't be storing invalid ips *)
+              | Error (`Msg msg) -> raise Api_errors.(Server_error (internal_error, [Printf.sprintf "could not parse ip address (%s), msg: %s" ip_str msg]))
+            in
+            let pool_host_ips =
+               Xapi_pool_helpers.get_master_slaves_list ~__context |>
+               List.map (fun x -> Db.Host.get_address ~__context ~self:x) |>
+               List.map parse_ip
+            in
+            let ensure_same_protocol ips =
+              let netmask = pifrec.API.pIF_netmask |> parse_ip in
```

Comment:
What PIF is this `pifrec` from? It is one of the host management PIFs, i.e. one of PIFs that have their IP addresses represented in the `host.address` fields that are used in `pool_host_ips`?

Reply:
are you saying I should use `master --> netmask instead of pifrec --> netmask`?

the pif comes from the command line

Reply:
What is the network that the clustering system is using? If it is always the management network of the pool (as this PR is assuming), then why do we get another PIF from the API call? What it is for?

Reply:
it appears that the only purpose of the PIF is to provide the `local_ip` address for the `init_config` which is then passed to clusterd - so this is clearly the right network. I'm not sure if getting the addresses from the master/slaves list is correct though

Reply:
@edwintorok do you know if xapi knows about the IP addresses of the cluster network? perhaps the check instead needs to be made in xapi-clusterd